src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt

```diff
@@ -1105,27 +1106,35 @@ sealed class RsDiagnostic(
             )
     }
 
-    class MissingFieldsInTuplePattern(pat: RsPat, private val declaration: RsFieldsOwner) : RsDiagnostic(pat) {
+    class MissingFieldsInTuplePattern(
+        pat: RsPat,
+        private val declaration: RsFieldsOwner,
+        private val expectedAmount: Int,
+        private val actualAmount: Int
+    ) : RsDiagnostic(pat) {
         override fun prepare(): PreparedAnnotation {
             val itemType = if (declaration is RsEnumVariant) "Enum variant" else "Tuple struct"
             return PreparedAnnotation(
                 ERROR,
                 E0023,
-                // TODO: provide number of expected/actual fields
-                "$itemType pattern does not correspond to its declaration",
+                "$itemType pattern does not correspond to its declaration: expected $expectedAmount ${pluralize("field", expectedAmount)}, found $actualAmount",
                 fixes = listOf(AddStructFieldsPatFix(element))
             )
         }
     }
 
-    class MissingFieldsInStructPattern(pat: RsPat, private val declaration: RsFieldsOwner) : RsDiagnostic(pat) {
+    class MissingFieldsInStructPattern(
+        pat: RsPat,
+        private val declaration: RsFieldsOwner,
+        private val missingFields: List<RsFieldDecl>
+    ) : RsDiagnostic(pat) {
         override fun prepare(): PreparedAnnotation {
             val itemType = if (declaration is RsEnumVariant) "Enum variant" else "Struct"
+            val missingFieldNames = escapeString(missingFields.joinToString(", ") { "`${it.name!!}`" })
             return PreparedAnnotation(
                 ERROR,
                 E0027,
-                // TODO: provide name of missing fields
-                "$itemType pattern does not correspond to its declaration",
+                "$itemType pattern does not mention following ${pluralize("field", missingFields.size)}: $missingFieldNames",
```

Comment:
```suggestion
                "$itemType pattern does not mention the following ${pluralize("field", missingFields.size)}: $missingFieldNames",
```

Reply:
Or drop `following` at all. It will simplify error message and will be the same as the compiler message